include/crab/domains/object_domain.hpp

```diff
@@ -0,0 +1,3312 @@
+    CRAB_LOG("object", crab::outs() << "Result6=" << res << "\n";);
+
+    res &= (left.m_uf_regs_dom <= right.m_uf_regs_dom);
+    CRAB_LOG("object", crab::outs() << "Result7=" << res << "\n";);
+    return res;
+  }
+
+  bool is_unknown_region(const variable_t &v) {
+    return v.get_type().is_unknown_region();
+  }
+
+  static void ERROR_IF_NOT_REGION(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_region()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not a region at line ", line);
+    }
+  }
+
+  static void ERROR_IF_ARRAY_REGION(const variable_t &v, unsigned line) {
+    if (v.get_type().is_array_region()) {
+      CRAB_ERROR(v, ":", v.get_type(), " cannot contain an array at line ",
+                 line);
+    }
+  }
+
+  static void ERROR_IF_NOT_REF(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_reference()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not a reference at line ", line);
+    }
+  }
+
+  static void ERROR_IF_NOT_INT(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_integer()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not an integer at line ", line);
+    }
+  }
+
+  /***************** Fields and Object id operations *****************/
+  boost::optional<obj_id_t> get_obj_id(const variable_t &rgn) {
+    if (!m_flds_id_map) {
+      return boost::none;
+    }
+    auto it = (*m_flds_id_map).find(rgn);
+    if (it == (*m_flds_id_map).end()) {
+      return boost::none;
+    }
+    return it->second;
+  }
+
+  obj_id_t get_obj_id_or_fail(variable_t rgn) {
+    assert(m_flds_id_map);
+    auto it = (*m_flds_id_map).find(rgn);
+    if (it == (*m_flds_id_map).end()) {
+      CRAB_ERROR(domain_name(),
+                 "::get_odi, the odi map does not include the region ", rgn,
+                 " belonging to an abstract object");
+    }
+    return it->second;
+  }
+
+  void get_obj_flds(const obj_id_t &id, variable_vector_t &obj_flds) const {
+    assert(m_flds_id_map);
+    for (auto kv : (*m_flds_id_map)) {
+      if (kv.second == id) {
+        obj_flds.push_back(kv.first);
+      }
+    }
+  }
+
+  void get_obj_ghost_flds(const obj_id_t &id,
+                          ghost_variable_vector_t &obj_ghost_flds) const {
+    variable_vector_t obj_flds;
+    get_obj_flds(id, obj_flds);
+    for (auto v : obj_flds) {
+      auto gvars_opt = get_gvars(v);
+      obj_ghost_flds.push_back(*gvars_opt);
+    }
+  }
+
+  void get_obj_dom_flds(const obj_id_t &id,
+                        base_dom_variable_vector_t &obj_dom_flds) const {
+    ghost_variable_vector_t obj_ghost_flds;
+    get_obj_ghost_flds(id, obj_ghost_flds);
+    for (auto v : obj_ghost_flds) {
+      if (v.has_offset_and_size()) {
+        obj_dom_flds.push_back(v.get_offset_and_size().get_offset());
+        obj_dom_flds.push_back(v.get_offset_and_size().get_size());
+      }
+      obj_dom_flds.push_back(v.get_var());
+    }
+  }
+
+  void update_fields_id_map(const variable_t &rgn, const obj_id_t &id) {
```

Comment:
fields -> flds to be consistent?

Reply:
This function will be called if we evaluate either DSA intrinsic or region int statement in crabIR.